src/main/scala/org/constellation/rewards/RewardsManager.scala

```diff
@@ -0,0 +1,201 @@
+package org.constellation.rewards
+
+import cats.effect._
+import cats.effect.concurrent.Ref
+import cats.implicits._
+import org.constellation.checkpoint.CheckpointService
+import org.constellation.consensus.Snapshot
+import org.constellation.domain.observation.Observation
+import org.constellation.p2p.PeerNotification
+import org.constellation.primitives.Schema.{AddressCacheData, CheckpointEdge}
+import org.constellation.primitives.{ChannelMessage, Transaction}
+import org.constellation.rewards.RewardsManager.rewardDuringEpoch
+import org.constellation.storage.{AddressService, SnapshotBroadcastService}
+import org.constellation.trust.TrustEdge
+
+class RewardsManager[F[_]: Concurrent](
+  eigenTrust: EigenTrust[F],
+  checkpointService: CheckpointService[F],
+  addressService: AddressService[F],
+  snapshotBroadcastService: SnapshotBroadcastService[F],
+) {
+  final val rewards: Ref[F, Map[String, Double]] = Ref.unsafe(Map.empty)
+
+  def updateBySnapshot(snapshot: Snapshot): F[Unit] =
+    for {
+      observations <- observationsFromSnapshot(snapshot)
+      _ <- eigenTrust.retrain(observations)
+
+      snapshotNum <- getSnapshotNumber
+
+      hashesTrustMap <- eigenTrust.getTrustForAddressHashes
+
+      distribution = RewardsManager
+        .rewardDistribution(hashesTrustMap.keys.toSeq, hashesTrustMap)
+        .mapValues(_ * rewardDuringEpoch(snapshotNum))
+
+      _ <- updateRewards(distribution)
+      _ <- updateAddressBalances(distribution)
+    } yield ()
+
+  // TODO: Check if this is the correct way of checking the snapshot number
+  // Assumption that `getRecentSnapshots` returns all snapshots may be failure
+  private def getSnapshotNumber: F[Int] =
+    snapshotBroadcastService.getRecentSnapshots.map(_.size)
+
+  private def observationsFromSnapshot(snapshot: Snapshot): F[Seq[Observation]] =
+    snapshot.checkpointBlocks.toList
+      .traverse(checkpointService.fullData)
+      .map(_.flatten.flatMap(_.checkpointBlock.observations))
+
+  private def updateRewards(distribution: Map[String, Double]): F[Unit] =
+    rewards.modify(prevDistribution => (prevDistribution |+| distribution, ()))
+
+  /*
+    TODO: Not sure if it is the correct way of updating balances by rewards.
+    1. putUnsafe takes Account.hash but we pass there Id.address which could differ. Maybe we should do Address(id.address).hash.
```

Comment:
That one is tricky @buckysballs. I decided to go for `lookup` + `putUnsafe` for sending the rewards. Otherwise I would need to create a transaction but not sure what should I put as `src` address there 🤔.

Let me know what you think :)

Reply:
You'll want to use a lock here, so I'd make a new method on AddressService calling memPool.update. Also, Id.address is the correct address for node Id.